onnx/onnx_backend.h

```diff
@@ -0,0 +1,914 @@
+ * @retval ONNX_STATUS_UNSUPPORTED_PARAMETER The function call failed because
+ *                                           the backend does not support the
+ *                                           particular data type, memory type,
+ *                                           or shape specified in one of the
+ *                                           operators.
+ * @retval ONNX_STATUS_UNSUPPORTED_SHAPE The function call failed because the
+ *                                       backend does can not support the
+ *                                       shape of input or output tensors.
+ *                                       This error code may be returned when
+ *                                       the backend supports variable-size
+ *                                       inputs and outputs, and the tensor
+ *                                       shape was provided to the graph as a
+ *                                       symbolic variable.
+ * @retval ONNX_STATUS_UNIDENTIFIED_NAME The function call failed because one
+ *                                       of the ValueInfoProto.name value in
+ *                                       ModelProto.graph.input or
+ *                                       ModelProto.graph.output doesn't have a
+ *                                       match in the inputDescriptors or
+ *                                       outputDescriptors.
+ * @retval ONNX_STATUS_NO_SYSTEM_MEMORY The function call failed because the
+ *                                      backend could not allocate enough
+ *                                      system memory to parse, analyze, and
+ *                                      initialize the tensor locations.
+ * @retval ONNX_STATUS_NO_SYSTEM_RESOURCES The function call failed due to
+ *                                         insufficient non-memory system
+ *                                         resources (e.g. file handles) to
+ *                                         initialize the tensor locations.
+ * @retval ONNX_STATUS_NO_DEVICE_MEMORY The function call failed due to
+ *                                      insufficient backend-specific memory to
+ *                                      initialize the tensor locations.
+ * @retval ONNX_STATUS_NO_DEVICE_RESOURCES The function call failed due to
+ *                                         insufficient non-memory
+ *                                         backend-specific resources (e.g.
+ *                                         command queues) to initialize the
+ *                                         tensor locations.
+ * @retval ONNX_STATUS_INTERNAL_ERROR The function call failed because the
+ *                                    backend experienced an unrecovered
+ *                                    internal error.
+ */
+ONNX_PUBLIC onnxStatus ONNX_ABI ONNX_SYMBOL_NAME(onnxSetGraphIO)(
+    onnxGraph graph,
+    uint32_t inputsCount,
+    const onnxTensorDescriptor* inputDescriptors,
```

Comment:
I think it will be more efficient to make parameter-passing positional instead of name-based. This is going to be invoked for every inference. In particular, it may be worth eliminating the name from onnxTensorDescriptor.

Reply:
SetGraphIO is actually intended to be called once at setup and you can invoke RunGraph multiple times reusing buffers if you want to. It matches well devices with costly memory allocation.

But I agree that it's not perfect solution for frameworks that do dynamic memory allocation (if it's on cpu especially). @Maratyszcza - thoughts on positional arguments?

Reply:
Passing parameters by name is convenient, and I expect that the time spent on string matching would be unnoticeable, despite its contribution is dependent on the use-case:
- In graphs with static tensor shapes, `onnxSetGraphIO` would be called only once, followed by many calls to `onnxRunGraph`. The cost of `onnxSetGraphIO`, including the cost of string matching, will be amortized across many inference invocations, and thus negligible.
- In graph with dynamic tensor shapes, `onnxSetGraphIO` would be called for every, or nearly every, call to `onnxRunGraph`. In each call `onnxSetGraphIO` would have to do shape inference for all internal tensors, memory allocation for internal blobs, and potentially code generation. String matching doesn't add much time on top of the above issues.

In the current models, it is typical to have very few (< 10) input and output tensors. If in the future we'd have use-cases where `onnxSetGraphIO` is called for every inference invocation, and the number of input and output tensors is so big that name matching becomes an issue, backend libraries could implement perfect hashing to tensor names, as the names of all inputs and outputs are available when graph is constructed.

Reply:
@Maratyszcza , this is really a great detail clarification. I understood and agreed on these points since the similar questions are also there in our side. However, I personally still think that "Node" is a better granularity to check the capability and do execution. Probably we should fix the issue of "Node" in onnx now. Thoughts?

Reply:
@linkerzhang It seems that your comment is for another thread, about the `onnxGetBackendCompatibility` function. We don't have any use-cases where `ModelProto` passed to `onnxGetBackendCompatibility` need to contain more than one Operator. I suggest we change the documentation to require that a user always pass `ModelProto` with an only Operator in the graph.